docs/rfc/rfc-007-deterministic-proto-bytes.md

```diff
@@ -0,0 +1,104 @@
+of messages within a single binary, these guarantees are not good enough for our
+use case in Tendermint. We require multiple different versions of a binary running
+Tendermint to be able to inter-operate. Additionally, we require that multiple different
+systems written in _different languages_ be able to participate in different aspects
+of the protocols of Tendermint and be able to verify the integrity of the messages
+they each produce.
+
+While this has not yet created a problem that we know of in a running network, we should
+make sure to provide stronger guarantees around the serialized representation of the messages
+used within the Tendermint consensus algorithm to prevent any issue from occurring.
+
+
+## Discussion
+
+Proto has the following points of variability that can produce non-deterministic byte representation:
+
+1. Encoding order of fields within a message.
+
+Proto allows fields to be encoded in any order and even be repeated.
+
+2. Encoding order of elements of a repeated field.
+
+`repeated` fields in a proto message can be serialized in any order.
+
+3. Presence or absence of default values.
+
+Types in proto have defined default values similar to Go's zero values. 
+Writing or omitting a default value are both legal ways of encoding a wire message.
+
+4. Serialization of 'unknown' fields. 
+
+Unknown fields can be present when a message is created by a binary with a newer 
+version of the proto that contains fields that the deserializer in a different 
+binary does not yet know about. Deserializers in binaries that do not know about the field
+will maintain the bytes of the unknown field but not place them into the deserialized structure.
+
+We have a few options to consider when producing this stable representation.
+
+### Use only compliant serializers and constrain field usage
+
+According to [Cosmos-SDK ADR-27][cosmos-sdk-adr-27], when message types obey a simple 
+set of rules, gogoproto produces a consistent byte representation of serialized messages.
+This seems promising, although more research is needed to guarantee gogoproto always
+produces a consistent set of bytes on serialized messages. This would solve the problem 
+within Tendermint as written in Go, but would require ensuring that there are similar
+serializers written in other languages that produce the same output as gogoproto.
+
+### Reorder serialized bytes to ensure determinism.
+
+The serialized form of a proto message can be transformed into a canonical representation
+by applying simple rules to the serialized bytes. Re-ordering the serialized bytes
+would allow Tendermint to produce a canonical byte representation without having to 
+simultaneously maintain a custom proto marshaller. 
+
+This could be implemented as a function in many languages that performed the following steps:
+
+1. Reordered all fields to be in tag-sorted order.
+2. Reordered all `repeated` sub-fields to be in lexicographically sorted order.
```

Comment:
Q: Is this already the case? I know that for instance ValidatorSets are sorted already but is e.g. is Evidence lexicographically sorted?

Reply:
There’s no guarantee that sorting the field before encoding affects the encoding. In Go it does, and in practice it’s a sensible assumption. But it could change at any time.